### PR TITLE
feat(ui): replicaset children of deployment should sort by revision (#4249)

### LINKED ISSUE #4249: replicaset children of deployment should sort by revision
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

Right now, replicaset children of deployments are sorted alphabetically by replicaset name.
Each replicaset has a revision (`rev`), it would be immensely helpful if they were sorted by that before sorting by name.

**To Reproduce**

1. Have a deployment with a replicaset with multiple revisions
1. Publish a bunch of new revisions
1. Look at graph and see things in a fairly randomized order -- namely they are sorted by replicaset name and not rev

**Expected behavior**

Sort first by `rev` and then by `name`

**Screenshots**

image

**Version**

v1.6.1+159674e

**Logs**

### PR DESCRIPTION
**Pull request by jsoref**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Optional. My organization is added to USERS.md.
* [x] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)). 

closes #4249

### CODE REVIEW

File: ui/src/app/applications/components/application-resource-tree/application-resource-tree.test.tsx [hunk 1]
```diff
@@ -21,4 +21,51 @@ Namespace: my-ns
 Name: my-name
 Images:
 - my-image:v1`)
-});
\ No newline at end of file
+});
+
+/*
+test("compareNodes", () => {
```
Comment (jsoref) [inline ui/src/app/applications/components/application-resource-tree/application-resource-tree.test.tsx:27]: `compareNodes` isn't an exported function, but to test the function one can add `export` to it and then add it to the imports and then uncomment this code.

File: ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx [hunk 2]
```diff
@@ -116,7 +116,22 @@ function filterGraph(app: models.Application, filteredIndicatorParent: string, g
...
+        const numberA = Number(a);
+        const numberB = Number(b);
+        if (isNaN(numberA) || isNaN(numberB)) {
+            return a.localeCompare(b);
+        }
+        return Math.sign(numberA - numberB);
+    }
+    return (
+        orphanedToInt(first.orphaned) - orphanedToInt(second.orphaned) ||
+        compareRevision(first.resourceVersion, second.resourceVersion) ||
```
Comment (alexmt) [inline ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx:132]: I think we want to sort resources of same group/kind by revision independently and latest revision should go first:
 
```
Deployment rev 2
Deployment rev 1
Rollout rev 2
Rollout rev 1
```

instead of 

```
Deployment rev 2
Rollout rev 2
Deployment rev 1
Rollout rev 1
```

In this case we need to switch `compareRevision(first.resourceVersion, second.resourceVersion)` and `nodeKey(first).localeCompare(nodeKey(second))`. 

```
        orphanedToInt(first.orphaned) - orphanedToInt(second.orphaned) ||
        nodeKey(first).localeCompare(nodeKey(second)) ||
        compareRevision(first.resourceVersion, second.resourceVersion)
```

Comment (alexmt) [inline ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx:132]: The revision is stored in `info` field, not in `resourceVersion`. Also it is available only for some resources so we need to check if `info` has it.

https://github.com/argoproj/argo-cd/blob/9f47a116219700a46750707fd4d0e5f835404444/ui/src/app/shared/models.ts#L324

Here is sample resource:

```json
{
    "group": "apps",
    "version": "v1",
    "kind": "Deployment",
    "namespace": "argocd",
    "name": "argocd-redis-ha-haproxy",
    "uid": "6492f1e5-3301-4d02-adca-7c6a63109840",
    "info": [
        {
            "name": "Revision",
            "value": "Rev:1"
        }
    ],
    "resourceVersion": "319468248",
    "health": {
        "status": "Healthy"
    },
    "createdAt": "2020-06-10T00:54:59Z"
}
```

File: ui/src/app/applications/components/application-resource-tree/application-resource-tree.test.tsx (hunk 1 above)
Comment (alexmt) [reply ui/src/app/applications/components/application-resource-tree/application-resource-tree.test.tsx:27]: Thanks for prepare test! Please export function and uncomment the test.

**Review by alexmt — CHANGES_REQUESTED** [review]

File: ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx [hunk 3]
```diff
@@ -115,8 +115,30 @@ function filterGraph(app: models.Application, filteredIndicatorParent: string, g
...
+    function compareRevision(a: string, b: string) {
+        const numberA = Number(a);
+        const numberB = Number(b);
+        if (isNaN(numberA) || isNaN(numberB)) {
+            return a.localeCompare(b);
+        }
+        return Math.sign(numberA - numberB);
+    }
+    function getRevision(a: ResourceTreeNode) {
+        const value = a.info.filter(b=>b.name=="Revision" && b)[0]?.value;
```
Comment (jsoref) [inline ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx:131]: I'm not a typescript expert... And this is fairly magical. I think it's right. It could be totally wrong, it's definitely a bit ugly. Happy to reformat.

File: ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx [hunk 4]
```diff
@@ -115,8 +115,30 @@ function filterGraph(app: models.Application, filteredIndicatorParent: string, g
...
+        if (!value)
+            return 0;
+        return value.replace(/^Rev:/,"");
```
Comment (jsoref) [inline ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx:134]: The example given included a `Rev:` prefix. I'm assuming that it's always there, but if it isn't, this will still do _something_.

File: ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx [hunk 5]
```diff
@@ -115,8 +115,30 @@ function filterGraph(app: models.Application, filteredIndicatorParent: string, g
...
+    }
+    return (
+        orphanedToInt(first.orphaned) - orphanedToInt(second.orphaned) ||
+        nodeKey(first).localeCompare(nodeKey(second)) ||
+        compareRevision(getRevision(first), getRevision(second)) ||
```
Comment (jsoref) [inline ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx:139]: I swapped the order

File: ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx [hunk 6]
```diff
@@ -115,8 +115,30 @@ function filterGraph(app: models.Application, filteredIndicatorParent: string, g
...
+        0
```
Comment (jsoref) [inline ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx:140]: This makes it easier to add additional rules above w/o taking blame for unrelated lines.

**Review by jsoref — COMMENTED** [review]
For the time being, I'm appending commits showing the logical steps I've taken. I'm happy to squash/rebase.

This round isn't very tested.

But I did use: [typescriptlang playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgEoQM4HsCuUnozIDeAUMhcjsACYBcyGYUoA5gNzmUDWo9jzNp0rIQcALaYADoggMmLEBy4Uxk+YKXDKAN2gZgWEBsXKRrKLikmhpAL6lQkWLLSZc+CADkIYAO5YULxKAJIgMFgkKshgcFCsvgAycABGEAA2GAzEANpqcgKmALo2SnbaFLHxvoRZbth4BBAwOUUVyOmpGXW5+aWsJYVs5dFsUJh1iVhwNABCcJ0gSFBhFhOt7RAAHs5i6QCqqIl1Cmwb9o7g0PBIyFMz84vLq+MYGFEiABZYTH1DWqNrP9lA4nNdXOgGp4vFgaChtpAQDR3pCPE0iGQRDJxuBagxUY0IIRziJQBFsn9TlpkDoFjgClTWHYSZQQL4AkE2GEIgB+fHuQk+fyBYKsblYdrAcRwBIYPnAlkUV5oiAANX0hmMwPaCHGcEgNAAgmB5YzOKCri5bgTPAAVcbeWHwnYQJEogXQp0fShMfU4OX9dqfCALMCfU2aMyUb5YbjylJYLDpEMgdqWLAm-lQpD2iCOuFpiAAR2orwAClAcCA2PHE8m4KnooEpJ8GxAaLWkynzaQYFWEGBNcgEmB0DpgAYjAAKOBZlW5-MQACU3ooCCMTBpdJQAF5kHAAHRkrAHmDAdLOKcpHcAPhSB-yO53ACIxxPNc-kAAyL-IFJLnIAAYih5A9aXSelJSIKcAEJwPpJdohEcYwDwEBkEAwtUKgdD4IgA9xikTokCnAB6AA9Mc6FIgAaZ9nyXHt1xAbBkwPdIsFYKdSBHN9JxAKdiGPOgcmIPpnzgPiPxovC6GfAAmZ87BosSJDkV8IHHfjnxk7c5Pkg8AGYlKKGjsVdUdmiyVoaOVQl1Sgfi5J06h6HomjgjcnTxJ0598gwGQkGfZyZI1IwQosKxnLsRClyAA) -- The TS components are slightly fuzzed, as I didn't fill in some dependent components that I wasn't using.

File: ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx (hunk 3 above)
Comment (jsoref) [reply ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx]: This requires ts3.7, argo is currently using 3.5. We can upgrade argo first in general and revisit that as a whole.

File: ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx (hunk 2 above)
Comment (jsoref) [reply ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx]: @alexmt: I think this is now implemented, but I'm not sure I can construct a proper test for this, can you provide sample data?

**Review by alexmt — APPROVED** [review]
LGTM. Thank you @jsoref !
